source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -0,0 +1,528 @@
+  ASSERT(connect_finished_ || connections_[0]->connecting());
+  return connections_[0]->connecting();
+}
+
+void HappyEyeballsConnectionImpl::write(Buffer::Instance& data, bool end_stream) {
+  if (connect_finished_) {
+    connections_[0]->write(data, end_stream);
+    return;
+  }
+
+  post_connect_state_.write_buffer_ = dispatcher_.getWatermarkFactory().createBuffer(
+      []() -> void { ASSERT(false); }, [this]() -> void { this->onWriteBufferHighWatermark(); },
+      []() -> void { ASSERT(false); });
+  if (per_connection_state_.buffer_limits_.has_value()) {
+    post_connect_state_.write_buffer_.value()->setWatermarks(
+        per_connection_state_.buffer_limits_.value());
+  }
+  post_connect_state_.write_buffer_.value()->move(data);
+  post_connect_state_.end_stream_ = end_stream;
+}
+
+void HappyEyeballsConnectionImpl::setBufferLimits(uint32_t limit) {
+  if (!connect_finished_) {
+    ASSERT(!per_connection_state_.buffer_limits_.has_value());
+    per_connection_state_.buffer_limits_ = limit;
+    if (post_connect_state_.write_buffer_.has_value()) {
+      post_connect_state_.write_buffer_.value()->setWatermarks(
+          per_connection_state_.buffer_limits_.value());
+    }
+  }
+  for (auto& connection : connections_) {
+    connection->setBufferLimits(limit);
+  }
+}
+
+uint32_t HappyEyeballsConnectionImpl::bufferLimit() const {
+  if (!connect_finished_) {
+    if (per_connection_state_.buffer_limits_.has_value()) {
+      return per_connection_state_.buffer_limits_.value();
+    }
+    return 0;
+  }
+  return connections_[0]->bufferLimit();
+}
+
+bool HappyEyeballsConnectionImpl::aboveHighWatermark() const {
+  if (!connect_finished_) {
+    return above_write_high_water_mark_;
+  }
+
+  return connections_[0]->aboveHighWatermark();
+}
+
+const ConnectionSocket::OptionsSharedPtr& HappyEyeballsConnectionImpl::socketOptions() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->socketOptions();
+}
+
+absl::string_view HappyEyeballsConnectionImpl::requestedServerName() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->requestedServerName();
+}
+
+StreamInfo::StreamInfo& HappyEyeballsConnectionImpl::streamInfo() {
+  // Note, this might change before connect finishes.
+  return connections_[0]->streamInfo();
+}
+
+const StreamInfo::StreamInfo& HappyEyeballsConnectionImpl::streamInfo() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->streamInfo();
+}
+
+absl::string_view HappyEyeballsConnectionImpl::transportFailureReason() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->transportFailureReason();
+}
+
+bool HappyEyeballsConnectionImpl::startSecureTransport() {
+  if (!connect_finished_) {
+    per_connection_state_.start_secure_transport_ = true;
+  }
+  bool ret = false;
+  for (auto& connection : connections_) {
+    ret = connection->startSecureTransport();
```

Comment:
returning the result of the last, rather than combined result?

Reply:
Oh, good point. (I *think* it won't matter in practice because the impl of this method returns true or false based on the type of connection not based on anything dynamic. But still good point! :>)